lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,959 @@
-      // does not work because: (1) interfaces don't have a defined way to get
-      // their bit width and (2) interfaces have a symbol table that is used to
-      // verify internal ops, but this requires looking arbitrarily far upwards
-      // to find other symbols.
-      builder.create<sv::VerbatimOp>(loc, name + " " + name + "();");
-      break;
-    case ElementInfo::Vector: {
-      auto type = hw::UnpackedArrayType::get(builder.getIntegerType(info.width),
-                                             info.depth);
-      builder.create<sv::InterfaceSignalOp>(loc, name, type);
-      break;
+    // Error if a matching parent or companion do not exist.
+    if (parentIDMap.count(bundle.getID()) == 0) {
+      emitCircuitError() << "no parent found with 'id' value '"
+                         << bundle.getID().getValue().getZExtValue() << "'\n";
+      removalError = true;
+      return false;
     }
-    case ElementInfo::Ground: {
-      auto type = builder.getIntegerType(info.width);
-      builder.create<sv::InterfaceSignalOp>(loc, name, type);
-      break;
+    if (companionIDMap.count(bundle.getID()) == 0) {
+      emitCircuitError() << "no companion found with 'id' value '"
+                         << bundle.getID().getValue().getZExtValue() << "'\n";
+      removalError = true;
+      return false;
     }
-    case ElementInfo::String:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported string type>;");
-      break;
-    case ElementInfo::Boolean:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported boolean type>;");
-      break;
-    case ElementInfo::Integer:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported integer type>;");
-      break;
-    case ElementInfo::Double:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported double type>;");
-      break;
-    case ElementInfo::Error:
-      llvm_unreachable("Shouldn't be here");
-      break;
+
+    // Recursively walk the AugmentedBundleType to generate interfaces and XMRs.
+    // Error out if this returns None.
+    auto iface = traverseBundle(bundle, bundle.getID(),
+                                companionIDMap.lookup(bundle.getID()).name);
+    if (!iface) {
+      removalError = true;
+      return false;
     }
-  }
 
-  interfaces.clear();
-  interfaceMap.clear();
-  interfaceKeys.clear();
+    // Instantiate the interface inside the parent.
+    builder.setInsertionPointToEnd(
+        parentIDMap.lookup(bundle.getID()).second.getBodyBlock());
+    auto symbolName = circuitNamespaceRef->newName(
+        ("__" + companionIDMap.lookup(bundle.getID()).name + "_" +
+         bundle.getDefName().getValue() + "__"));
```

Comment:
Sweet uniquification!

Reply:
Yep. We also need to get this into the `LowerToHW` symbol names, e.g., [her](https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L1764).

Reply:
GCT data taps could possibly also benefit from this. Almost sounds like a general analysis over the IR quite a few passes could reuse.